src/utils/dynamic-position.js

```diff
@@ -0,0 +1,78 @@
+
+export const ANCHOR_POSITION = {
+  top: {x: 0.5, y: 0},
+  'top-left': {x: 0, y: 0},
+  'top-right': {x: 1, y: 0},
+  bottom: {x: 0.5, y: 1},
+  'bottom-left': {x: 0, y: 1},
+  'bottom-right': {x: 1, y: 1},
+  left: {x: 0, y: 0.5},
+  right: {x: 1, y: 0.5}
+};
+
+/**
+ * Calculate the dynamic position for a popup to fit in a container.
+ * @param {Number} x - x position of the anchor on screen
+ * @param {Number} y - y position of the anchor on screen
+ * @param {Number} width - width of the container
+ * @param {Number} height - height of the container
+ * @param {Number} padding - extra space from the edge in pixels
+ * @param {Number} selfWidth - width of the popup
+ * @param {Number} selfHeight - height of the popup
+ * @param {String} anchor - type of the anchor, one of 'top', 'bottom',
+    'left', 'right', 'top-left', 'top-right', 'bottom-left' , and  'bottom-right'
+ * @returns {String} position - one of 'top', 'bottom',
+    'left', 'right', 'top-left', 'top-right', 'bottom-left' , and  'bottom-right'
+ */
+export function getDynamicPosition({
+  x, y,
+  width, height,
+  selfWidth, selfHeight,
+  anchor,
+  padding = 0
+}) {
+  let {x: anchorX, y: anchorY} = ANCHOR_POSITION[anchor];
+
+  let top = y - anchorY * selfHeight;
+  let bottom = top + selfHeight;
+  const yStep = 0.5;
+
+  if (top < padding) {
```

Comment:
There's a ton of logic and loops in this function, can we add comments along the way, maybe at every `if` clause to guide the reader on what's happening in terms of the UI (e.g. `prevent exceeding bottom edge` or something less terrible :p)?